### PR TITLE
Theme editor for feedback

### PR DESCRIPTION
**Pull request by kevinkoenig**

### CODE REVIEW

**Comment by rotorman** [comment]
In the new GUI SYSTEM -> Themes page, I can browse the themes on SD card, but cannot select (or edit) them anymore:
screenshot_tx16s_21-10-30_15-03-27

**Comment by kevinkoenig** [comment]
Use long press to bring up the menu that allows you to do things.  There are new pushes to the PR that you should take.

Sent from Mail for Windows

**Comment by rotorman** [comment]
Awesome, now I see them:
screenshot_tx16s_21-10-30_16-00-34

**Comment by pfeerick** [comment]
Speaking of screen orientation, you might want to add another `#define COLOR_PREVIEW_WIDTH` to `radio_theme.cpp` as you have it defined at the top for one orientation, but not the other ;)

Also, Martin (who has done some wonderful graphics for OTX and is doing some for ETX now also, as well as sharing some UI designs he thought of) proposed this layout... in particular, being able to put the colour selection in a combo box at the bottom, getting more width back to then pad out the colour bars ;)

image

**Comment by kevinkoenig** [comment]
I spoke to Jim about this design.  We thought having the color type picker as a choice would make it be two clicks to switch types.  That is why we didn’t go with it.  If the feedback says that is the way to go then I will change it.  I have no preference either way.

**Comment by rotorman** [comment]
Instead of a drop-down menu to pick HSB or RGB, there could be two buttons side-by-side, similarly to current design, where the current selection is highlighted:
https://user-images.githubusercontent.com/21011587/139536641-964b7932-119e-4f78-925f-95b8f01006c8.png
That could then save 1 click.

**Comment by JimB40** [comment]
Artifact in checks to check progress?

**Comment by rotorman** [comment]
> Artifact in checks to check progress?

I believe GH Actions is not triggered for a draft PR, thus no. @JimB40 I can build you a firmware though if you like. For TX16S?

**Comment by JimB40** [comment]
> Also, Martin (who has done some wonderful graphics for OTX and is doing some for ETX now also, as well as sharing some UI designs he thought of) proposed this layout... in particular, being able to put the colour selection in a combo box at the bottom, getting more width back to then pad out the colour bars. Something to think about ;)

1. 20-25px from bottom is hard touchable area, so mainly for info.
2. Dropdown needs two clicks instead of one.
3. There is enough space to make sliders wide enough not introduce errors
4. No white spaces.
5. Hex is shifted to left and now as label not editable field
6. HSV and RGB buttons to close to each other.
7. RGB sliders are reversed vertically (should be 0 on the bottom and 255 on the top)

Implementation differs from project but I think @kevinkoenig still works on that :)
Screen_Shot_2021-10-23_at_18 26 16

**Comment by JimB40** [comment]
> I believe GH Actions is not triggered for a draft PR, thus no. @JimB40 I can build you a firmware though if you like. For TX16S?

Yes please, better to have hands-on experience testing then screenshots

**Comment by rotorman** [comment]
@JimB40 Here you go (built with: `-DPCB=X10 -DPCBREV=TX16S -DDEFAULT_MODE=2 -DGVARS=YES -DPPM_UNIT=US -DLUA=YES -DCMAKE_BUILD_TYPE=Release ../` )
[20211031_ETX_TX16S_ThemeEditor.zip](https://github.com/EdgeTX/edgetx/files/7449146/20211031_ETX_TX16S_ThemeEditor.zip)

**Comment by pfeerick** [comment]
> 20-25px from bottom is hard touchable area, so mainly for info.

Not just the bottom... all edges, and more on some transmitters depending on the bezel. 

> Implementation differs from project but I think @kevinkoenig still works on that :) Screen_Shot_2021-10-23_at_18 26 16

So is this what you currently have in mind for it to finally look like? ^^^ If so, given there is a third button to make use of the space, I can see why a dropdown will added unnecessary clicks vs space used. And PAL was a colour palette rather than the slider bars, wasn't it?

**Comment by kevinkoenig** [comment]
> Speaking of screen orientation, you might want to add another `#define COLOR_PREVIEW_WIDTH` to `radio_theme.cpp` as you have it defined at the top for one orientation, but not the other ;)
> 
> Also, Martin (who has done some wonderful graphics for OTX and is doing some for ETX now also, as well as sharing some UI designs he thought of) proposed this layout... in particular, being able to put the colour selection in a combo box at the bottom, getting more width back to then pad out the colour bars. Something to think about ;)
> 
> image

also for NV14, there still needs to be some work done.  The color preview will need to be horizontal (which it supports), but color_preview_width wont make sense then.  It will be COLOR_PREVIEW_HEIGHT :).  Once we settle down on the screen design for wide screens I will finish narrow screens.

**Comment by kevinkoenig** [comment]
> > 20-25px from bottom is hard touchable area, so mainly for info.
> 
> Not just the bottom... all edges, and more on some transmitters depending on the bezel.
> 
> > Implementation differs from project but I think @kevinkoenig still works on that :) Screen_Shot_2021-10-23_at_18 26 16
> 
> So is this what you currently have in mind for it to finally look like? ^^^ If so, given there is a third button to make use of the space, I can see why a dropdown will added unnecessary clicks vs space used. And PAL was a colour palette rather than the slider bars, wasn't it?

please take a look at the current code.  It has removed PAL.  We might add it back.  I have also reworked a bunch of layout as per the feedback given so far.

**Comment by rotorman** [comment]
@kevinkoenig Thank you for the update!

I noticed that the default EdgeTX theme color blocks do not fit the vertical space:
screenshot_tx16s_21-11-01_11-34-27

whereas when browsing to the previous example themes by @JimB40, the color blocks are better aligned:
screenshot_tx16s_21-11-01_11-34-41

?

I still cannot delete the last theme in the list (here y2). When I try to delete it, it won't disappear from the list:
screenshot_tx16s_21-11-01_11-34-41

Here some impressions of preset state:
screenshot_tx16s_21-11-01_11-52-27

screenshot_tx16s_21-11-01_11-52-53

screenshot_tx16s_21-11-01_11-53-25

Attached also firmware for TX16S and Windows-Companion with TX16 simulator with the current state, for all who want to test this Draft-PR, but cannot build their own from GitHub.
Firmware for TX16S with analog sticks: 
[20211101_ETX_TX16S_themeeditor.zip](https://github.com/EdgeTX/edgetx/files/7452557/20211101_ETX_TX16S_themeeditor.zip)
Windows Companion with Simu & TX16S libsimu:
[WinCompanion_ThemeEditor_TX16S_simu.zip](https://github.com/EdgeTX/edgetx/files/7452553/WinCompanion_ThemeEditor_TX16S_simu.zip)

File: radio/src/gui/colorlcd/color_editor2.cpp [hunk 1]
```diff
@@ -0,0 +1,377 @@
...
+}
+
+/////////////////////////////////////////////////////////////////////////
+/////////////////////////////////////////////////////////////////////////
+/////////////////////////////////////////////////////////////////////////
+void setHexStr(StaticText* hexBox, uint32_t rgb)
+{
+  auto r = GET_RED(rgb), g = GET_GREEN(rgb), b = GET_BLUE(rgb);
+  char hexstr[80];
+  sprintf(hexstr, "%02X%02X%02X\n", r, g, b);
```
[the remark this answers is not in the document]
Comment (kevinkoenig) [reply radio/src/gui/colorlcd/color_editor2.cpp]: > r, g, b could be cast to unsigned int (see compiler warning):
> 
> ```
> color_editor2.cpp:183:23: warning: format '%X' expects argument of type 'unsigned int', but argument 3 has type 'long unsigned int' [-Wformat=]
>  sprintf(hexstr, "%02X%02X%02X\n", r, g, b);
>       |                    ~~~^             ~
>       |                       |             |
>       |                       unsigned int  long unsigned int
>       |                    %02lX
> ```

resolved.  issue comes when building for firmware (not simu)

**Comment by kevinkoenig** [comment]
> > After couple of seconds, the teaser disappears:
> 
> That's probably because the code optimistically assumes all the screenshshots are there, instead of look before load ;) If you have a theme with both the logo and screenshots present, it shows them in turn changing, every second or so.
> 
> I'm having some issues at times with `simu` generally atm - it doesn't seem to like the NV14 at times (bottom half of the simu screen keeps flashing the ETX logo instead of main screen, but once you're into a menu its fine... and I'm getting a `buffer overflow detected` crash from simu with part of this also... so it may be time to update something or clean some junk out...

this was a defect and should be resolved

**Comment by JimB40** [comment]
Stable now :) 

So there are four main areas of feedback:
- Navigation
- Layout and UI
- Bugs or Something not working as expected
- Efficiency

I think we should cover them 1 by 1. 
@kevinkoenig pick one you want to start with :)

**Comment by kevinkoenig** [comment]
> Stable now :)
> 
> So there are four main areas of feedback:
> 
> * Navigation
> * Layout and UI
> * Bugs or Something not working as expected
> * Efficiency
> 
> I think we should cover them 1 by 1. @kevinkoenig pick one you want to start with :)

why don't we start with navigation and layout.  I think we can do 2 at a time.  for bugs we can wait until libopenui gets merged which fixes a whole bunch of touch and slide issues.

**Comment by JimB40** [comment]
https://youtu.be/QR6KbDHJneY
Test abt navigation

**Comment by kevinkoenig** [comment]
> https://youtu.be/QR6KbDHJneY Test abt navigation

thanks jim for spending time and providing feedback.  I like most of it.  I want to discuss the control navigation that you propose.  I was trying to make these controls ListBase(), ListBox(), and ColorList(), ColorEditor() work exactly like NumberEdit and TextEdit() in a form page.  I hear you that this creates some confusion about navigation, but it is completely consistent with the way the other controls work.  Let's figure out the best way to get all of this done, as we are creating new UI concepts that have not been implemented ever before.

**Comment by mhotar** [comment]
Nice video. If I can put my two cents in. Top bar uses PAGE UP/DOWN to switch items. We can perhaps use it here as well by giving the items some representation there. As a side effect it gives to the color bars more space by moving HSV/RGB to the top.
themeedit

**Comment by kevinkoenig** [comment]
> > https://youtu.be/QR6KbDHJneY Test abt navigation
> 
> thanks jim for spending time and providing feedback. I like most of it. I want to discuss the control navigation that you propose. I was trying to make these controls ListBase(), ListBox(), and ColorList(), ColorEditor() work exactly like NumberEdit and TextEdit() in a form page. I hear you that this creates some confusion about navigation, but it is completely consistent with the way the other controls work. Let's figure out the best way to get all of this done, as we are creating new UI concepts that have not been implemented ever before.

So I did a lot of investigation on how we can make PAGE controls move focus between controls.  All I can say right now is that it will be very hard to make the standard controls work.  Specifically the navigation for controls like TextButton() do not support paging via hardware keys (PAGE). 

> Nice video. If I can put my two cents in. Top bar uses PAGE UP/DOWN to switch items. We can perhaps use it here as well by giving the items some representation there. As a side effect it gives to the color bars more space by moving HSV/RGB to the top. themeedit

in your suggestion, what does the items button do?

**Comment by kevinkoenig** [comment]
@JimB40 most of what you asked for has been incorporated into this push.  It requires a libopenui change for some of the PGUP and PGDOWN processing to work the way you wanted.  That PR has also been pushed.  There are still a couple of defects that needs to be corrected, but we can build an image for you and you can try it out again.

**Comment by JimB40** [comment]
> Nice video. If I can put my two cents in. Top bar uses PAGE UP/DOWN to switch items. We can perhaps use it here as well by giving the items some representation there. As a side effect it gives to the color bars more space by moving HSV/RGB to the top.
> themeedit

Yes @mhotar I was thinking about that solution. I mean HSV, RGB (and Pallette) are like TABs on the top bar.
Problem is that even those buttons look identical as SAVE they behave differentely. (Other UI type)
So TAB like buttons behaves that you can switch between them but they never get focused. Instead of that focus is set to the first field in selected tab and icon/button is just higlighted.
Then we introduced SAVE which is differen UI animal. it first needs to be focused then wait to user interaction (ENTER). Mixing those two types of UI elements in one row is always confusing.
I like idea of those modes bottons/icons as people are used to switch beetween tabs with PAGE UP/DN and it gives  more space for sliders. In theme selecting screen (2nd one) tabs would be COLORS and DETAILS while in color picker screen tabs would be RGB HSV PALETTE
So not to mix those  SAVE is not visible in TopBar.In such a case  used UX solution is that whenever user changes anything after EXIT is pressed system display pop-up asking if I he wants to save or cancel changes.
Last but not least color picker should be usable in other places like widget settings (then as pop-up) so we should have then different type of those icon placement

**Comment by mhotar** [comment]
> in your suggestion, what does the items button do? 

They are not meant as buttons. They are more like tabs - the same how the main bar with icons works.
What do you see at the first picture is "Items" view (may be not the best name for it) selected. Switching to "Details" shows your Theme detail view.
The question is what to do with Save. If it means Exit and Save the page, we can use either the most left top icon as is in the rest of radio (Edit Mix etc..) or introduce a new icon as shown in next pictures. This icon is accesible only by touch screen, for the hw control it uses RTN button to Close and Save page.
themeedit2
mixesedit2
For closing the top bar pages we can perhaps use some picto graphics at the header icon.

topbar2

I know it probably does not fit to your concept, as I do not know how big UI rework are you planning. :)
Anyway I enjoyed the time.

One more thing. Maybe you can use the same method for your User level feature you are planning.
mixesedit2ba

File: radio/src/gui/colorlcd/radio_theme.cpp [hunk 2]
```diff
@@ -0,0 +1,723 @@
...
+          nameText->setText(theme->getName());
+          fileCarosell->setFileNames(theme->getThemeImageFileNames());
+        }
+        currentTheme = value;
+      });
+  listBox->setActiveIndex(tp->getThemeIndex());
+  listBox->setTitle(STR_THEME + std::string("s"));
+  listBox->setLongPressHandler([=] (event_t event) {
+    auto menu = new Menu(listBox,false);
+
```
Comment (pfeerick) [inline radio/src/gui/colorlcd/radio_theme.cpp:588]: How about adding adding `listBox->getSelected() != tp->getThemeIndex()` check to hide STR_ACTIVATE line if trying to make active theme active?

File: radio/src/gui/colorlcd/radio_theme.cpp [hunk 3]
```diff
@@ -0,0 +1,723 @@
...
+          listBox->setNames(tp->getNames());
+          listBox->setSelected(currentTheme);
+        }
+      });
+    });
+
+    // you cant delete the default theme or the currently active theme
+    if (listBox->getSelected() != 0 && listBox->getSelected() != tp->getThemeIndex()) {
+      menu->addLine(STR_DELETE, [=] () {
+        if (confirmationDialog("Delete Theme: ", "Are you sure?")) {
```
Comment (pfeerick) [inline radio/src/gui/colorlcd/radio_theme.cpp:663]: Can we add the theme name to the msg so we know which theme is being deleted - fixes PBKAC syndrome ;)

**Comment by pfeerick** [comment]
As promised, build for TX16S including libopenUI PR WIP => [TX16S_0020b51.zip](https://github.com/EdgeTX/edgetx/files/7489911/TX16S_0020b51.zip)

Thanks for adding the 'active theme' dot... that works just fine! 
image

'Enter capturing focus' issues seem to be gone, RTN exiting back to the wrong place seem to be gone, focus issues in edit colour page are mostly gone...when you use the roller to scroll, as soon as the colour bars are selected, any further rolling only changes focus between colour bars. Using PGUP/PGDN takes you between HSV/RGB/colour bars/save and Details/Save/theme elements... looks good! :)

**Comment by kevinkoenig** [comment]
Good call on both fronts.  Thanks

Sent from Mail for Windows

**Comment by rotorman** [comment]
John just posted nice video about the present state:
https://www.youtube.com/watch?v=jhuO7X1AmTk

File: .gitmodules [hunk 4]
```diff
@@ -1,6 +1,6 @@
 [submodule "libopenui"]
 	path = radio/src/thirdparty/libopenui
-	url = https://github.com/EdgeTX/libopenui.git
+	url = https://github.com/kevinkoenig/libopenui.git
```
Comment (raphaelcoeffic) [inline .gitmodules:3]: @pfeerick this needs to be fixed first before merging ;-)

File: radio/src/gui/colorlcd/CMakeLists.txt [hunk 5]
```diff
@@ -98,10 +105,6 @@ if(PXX2 OR MULTIMODULE)
   add_gui_src(radio_spectrum_analyser.cpp)
 endif()
 
-if(PXX2)
-  add_gui_src(access_settings.cpp)
-endif()
```
Comment (raphaelcoeffic) [inline radio/src/gui/colorlcd/CMakeLists.txt:103]: This change needs to be removed

**Comment by raphaelcoeffic** [comment]
@kevinkoenig I cleaned-up your PR, but it seems I screwed some headers....

**Comment by pfeerick** [comment]
I gave this a quick try locally on TX16S, and seems to be working fine. 

One last point before we just merge the sucker... (unless you can see something broken)... I can see (I'm in dumb user mode atm, so only from actually using this) PAGE UP/DN and TELE are being used for stuff... that's cool. But when when I'm here

image

i.e. scrolling up and down for a color element to change, can I not use RTN - either short or long to get out of that screen... I need to use PAGE UP/DN and then I can do it (or "cheat" and use touch screen).

**Comment by kevinkoenig** [comment]
> I gave this a quick try locally on TX16S, and seems to be working fine (now).
> 
> One last point before we just merge the sucker... (unless you can see something broken in de-conflicting this)... I can see (I'm in dumb user mode atm, so only from actually using this) PAGE UP/DN and TELE are being used for stuff... that's cool. But when when I'm here
> 
> image
> 
> i.e. scrolling up and down for a color element to change, can I not use RTN - either short or long to get out of that screen... I need to use PAGE UP/DN and then I can do it (or "cheat" and use touch screen).
> 
> Oh, and I only just realised - I can't use touch to switch between RGB and HSV "buttons" - only PAGE key(s)?

1. that is a bug that I just introduced.  
2. I added touch support to the pagebuttons so you can click them

**Review by pfeerick — APPROVED** [review]
Thank you, thank you! :) I think Christmas has come early!

**Comment by rotorman** [comment]
@kevinkoenig could you please have a second look at the following line:
https://github.com/EdgeTX/edgetx/blob/a9abf028f74c91eb1dba8a60ac84ce96d4fcadee/radio/src/gui/colorlcd/listbox.cpp#L70

```
radio/src/gui/colorlcd/listbox.cpp:70:34: warning: comparison of integer expressions of different signedness: 'uint32_t' {aka 'long unsigned int'} and 'int32_t' {aka 'long int'} [-Wsign-compare]
   70 |     LcdFlags circleColor = index == selected ? COLOR_THEME_PRIMARY2 : COLOR_THEME_PRIMARY1;
      |                            ~~~~~~^~~~~~~~~~~
```
`index` and `circleColor` are defined as uint32_t, whereas `selected` is defined as int32_t.

Please consider instead:
```
LcdFlags circleColor = index == (uint32_t)(selected ? COLOR_THEME_PRIMARY2 : COLOR_THEME_PRIMARY1);
```

**Pull request merged**